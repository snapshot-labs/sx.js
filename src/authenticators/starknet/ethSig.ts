import { Call, CallData, shortString } from 'starknet';
import EthSigAuthenticatorAbi from './abis/EthSigAuthenticator.json';
import { getChoiceEnum } from '../../utils/starknet-enums';
import {
  Authenticator,
  Envelope,
  Propose,
  Vote,
  UpdateProposal,
  ProposeCallArgs,
  VoteCallArgs,
  UpdateProposalCallArgs
} from '../../types';

const callData = new CallData(EthSigAuthenticatorAbi);

export default function createEthSigAuthenticator(): Authenticator {
  return {
    type: 'ethSig',
    createProposeCall(envelope: Envelope<Propose>, args: ProposeCallArgs): Call {
      const { space, authenticator } = envelope.data;

      if (!envelope.signatureData?.signature) {
        throw new Error('signature is required for this authenticator');
      }

      if (!envelope.signatureData?.message) {
        throw new Error('message is required for this authenticator');
      }

      const [r, s, v] = envelope.signatureData.signature;

      const compiled = callData.compile('authenticate_propose', [
        r,
        s,
        v,
        space,
        args.author,
        shortString.splitLongString(args.metadataUri),
        {
          address: args.executionStrategy.address,
          params: args.executionStrategy.params
        },
        args.strategiesParams,
        envelope.signatureData.message.salt
      ]);

      return {
        contractAddress: authenticator,
        entrypoint: 'authenticate_propose',
        calldata: compiled
      };
    },
    createVoteCall(envelope: Envelope<Vote>, args: VoteCallArgs): Call {
      const { space, authenticator } = envelope.data;

      if (!envelope.signatureData?.signature) {
        throw new Error('signature is required for this authenticator');
      }

      if (!envelope.signatureData?.message) {
        throw new Error('message is required for this authenticator');
      }

      const [r, s, v] = envelope.signatureData.signature;

      const compiled = callData.compile('authenticate_vote', [
        r,
        s,
        v,
        space,
        args.voter,
        args.proposalId,
        getChoiceEnum(args.choice),
        args.votingStrategies.map(strategy => ({
          index: strategy.index,
          params: strategy.params
        })),
        shortString.splitLongString(args.metadataUri)
      ]);

      return {
        contractAddress: authenticator,
        entrypoint: 'authenticate_vote',
        calldata: compiled
      };
    },
    createUpdateProposalCall(
      envelope: Envelope<UpdateProposal>,
      args: UpdateProposalCallArgs
    ): Call {
      const { space, authenticator } = envelope.data;

      if (!envelope.signatureData?.signature) {
        throw new Error('signature is required for this authenticator');
      }

      if (!envelope.signatureData?.message) {
        throw new Error('message is required for this authenticator');
      }

      const [r, s, v] = envelope.signatureData.signature;

      const compiled = callData.compile('authenticate_update_proposal', [
        r,
        s,
        v,
        space,
        {
          address: args.author
        },
        args.proposalId,
        {
          address: args.executionStrategy.address,
          params: args.executionStrategy.params
        },
        shortString.splitLongString(args.metadataUri),
        envelope.signatureData.message.salt
      ]);

      return {
        contractAddress: authenticator,
        entrypoint: 'authenticate_update_proposal',
        calldata: compiled
      };
    }
  };
}
