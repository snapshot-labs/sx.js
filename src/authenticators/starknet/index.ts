import createVanillaAuthenticator from './vanilla';
import createEthSigAuthenticator from './ethSig';
import createEthTxAuthenticator from './ethTx';
import createStarkSigAuthenticator from './starkSig';
import createStarkTxAuthenticator from './starkTx';
import { hexPadLeft } from '../../utils/encoding';
import type { Authenticator, NetworkConfig } from '../../types';

export function getAuthenticator(
  address: string,
  networkConfig: NetworkConfig
): Authenticator | null {
  const authenticator = networkConfig.authenticators[hexPadLeft(address)];
  if (!authenticator) return null;

  if (authenticator.type === 'vanilla') {
    return createVanillaAuthenticator();
  }

  if (authenticator.type === 'ethSig') {
    return createEthSigAuthenticator();
  }

  if (authenticator.type === 'ethTx') {
    return createEthTxAuthenticator();
  }

  if (authenticator.type === 'starkSig') {
    return createStarkSigAuthenticator();
  }

  if (authenticator.type === 'starkTx') {
    return createStarkTxAuthenticator();
  }

  return null;
}
