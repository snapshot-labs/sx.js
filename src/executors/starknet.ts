import { createStarknetExecutionParams } from '../utils/encoding';
import type { Call } from 'starknet';

export default function createStarknetExecutor() {
  return {
    type: 'starknet',
    getExecutionData(executorAddress: string, calls: Call[]) {
      return {
        executor: executorAddress,
        executionParams: createStarknetExecutionParams(calls)
      };
    }
  };
}
