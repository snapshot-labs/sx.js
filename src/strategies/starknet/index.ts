import createVanillaStrategy from './vanilla';
import createMerkleWhitelistStrategy from './merkleWhitelist';
import createErc20VotesStrategy from './erc20Votes';
import { hexPadLeft } from '../../utils/encoding';
import type { Strategy, NetworkConfig } from '../../types';

export function getStrategy(address: string, networkConfig: NetworkConfig): Strategy | null {
  const strategy = networkConfig.strategies[hexPadLeft(address)];
  if (!strategy) return null;

  if (strategy.type === 'vanilla') {
    return createVanillaStrategy();
  }

  if (strategy.type === 'whitelist') {
    return createMerkleWhitelistStrategy();
  }

  if (strategy.type === 'erc20Votes') {
    return createErc20VotesStrategy();
  }

  return null;
}
