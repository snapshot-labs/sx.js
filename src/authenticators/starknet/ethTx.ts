import { Call, CallData } from 'starknet';
import EthTxAuthenticatorAbi from './abis/EthTxAuthenticator.json';
import {
  Authenticator,
  Envelope,
  Propose,
  Vote,
  UpdateProposal,
  ProposeCallArgs,
  VoteCallArgs,
  UpdateProposalCallArgs
} from '../../types';

const callData = new CallData(EthTxAuthenticatorAbi);

export default function createEthTxAuthenticator(): Authenticator {
  return {
    type: 'ethTx',
    createProposeCall(envelope: Envelope<Propose>, args: ProposeCallArgs): Call {
      const { space, authenticator } = envelope.data.message;

      const compiled = callData.compile('authenticate_propose', [
        space,
        args.author,
        {
          address: args.executionStrategy.address,
          params: [args.executionStrategy.params]
        },
        args.strategiesParams
      ]);

      return {
        contractAddress: authenticator,
        entrypoint: 'authenticate_propose',
        calldata: compiled
      };
    },
    createVoteCall(envelope: Envelope<Vote>, args: VoteCallArgs): Call {
      const { space, authenticator } = envelope.data.message;

      const compiled = callData.compile('authenticate_vote', [
        space,
        args.voter,
        args.proposalId,
        args.choice,
        args.votingStrategies.map(strategy => ({
          index: strategy.index,
          params: [strategy.params]
        }))
      ]);

      return {
        contractAddress: authenticator,
        entrypoint: 'authenticate_vote',
        calldata: compiled
      };
    },
    createUpdateProposalCall(
      envelope: Envelope<UpdateProposal>,
      args: UpdateProposalCallArgs
    ): Call {
      const { space, authenticator } = envelope.data.message;

      const compiled = callData.compile('authenticate_update_proposal', [
        space,
        args.author,
        args.proposalId,
        {
          address: args.executionStrategy.address,
          params: [args.executionStrategy.params]
        }
      ]);

      return {
        contractAddress: authenticator,
        entrypoint: 'authenticate_update_proposal',
        calldata: compiled
      };
    }
  };
}
