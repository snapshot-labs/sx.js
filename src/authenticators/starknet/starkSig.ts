import { Call, CallData, shortString, uint256 } from 'starknet';
import StarkSigAuthenticatorAbi from './abis/StarkSigAuthenticator.json';
import { getChoiceEnum } from '../../utils/starknet-enums';
import {
  Authenticator,
  Envelope,
  Propose,
  Vote,
  UpdateProposal,
  ProposeCallArgs,
  VoteCallArgs,
  UpdateProposalCallArgs
} from '../../types';

const callData = new CallData(StarkSigAuthenticatorAbi);

export default function createStarkSigAuthenticator(): Authenticator {
  return {
    type: 'starkSig',
    createProposeCall(envelope: Envelope<Propose>, args: ProposeCallArgs): Call {
      const { authenticator } = envelope.data;

      if (!envelope.signatureData?.signature) {
        throw new Error('signature is required for this authenticator');
      }

      if (!envelope.signatureData?.message) {
        throw new Error('message is required for this authenticator');
      }

      const compiled = callData.compile('authenticate_propose', [
        envelope.signatureData.signature,
        envelope.data.space,
        envelope.signatureData.address,
        shortString.splitLongString(args.metadataUri),
        {
          address: args.executionStrategy.address,
          params: args.executionStrategy.params
        },
        args.strategiesParams,
        envelope.signatureData.message.salt
      ]);

      return {
        contractAddress: authenticator,
        entrypoint: 'authenticate_propose',
        calldata: compiled
      };
    },
    createVoteCall(envelope: Envelope<Vote>, args: VoteCallArgs): Call {
      const { authenticator } = envelope.data;

      if (!envelope.signatureData?.signature) {
        throw new Error('signature is required for this authenticator');
      }

      if (!envelope.signatureData?.message) {
        throw new Error('message is required for this authenticator');
      }

      const compiled = callData.compile('authenticate_vote', [
        envelope.signatureData.signature,
        envelope.data.space,
        envelope.signatureData.address,
        uint256.bnToUint256(args.proposalId),
        getChoiceEnum(args.choice),
        args.votingStrategies.map(strategy => ({
          index: strategy.index,
          params: strategy.params
        })),
        shortString.splitLongString(args.metadataUri)
      ]);

      return {
        contractAddress: authenticator,
        entrypoint: 'authenticate_vote',
        calldata: compiled
      };
    },
    createUpdateProposalCall(
      envelope: Envelope<UpdateProposal>,
      args: UpdateProposalCallArgs
    ): Call {
      const { authenticator } = envelope.data;

      if (!envelope.signatureData?.signature) {
        throw new Error('signature is required for this authenticator');
      }

      if (!envelope.signatureData?.message) {
        throw new Error('message is required for this authenticator');
      }

      const compiled = callData.compile('authenticate_update_proposal', [
        envelope.signatureData.signature,
        envelope.data.space,
        envelope.signatureData.address,
        uint256.bnToUint256(args.proposalId),
        {
          address: args.executionStrategy.address,
          params: args.executionStrategy.params
        },
        shortString.splitLongString(args.metadataUri),
        envelope.signatureData.message.salt
      ]);

      return {
        contractAddress: authenticator,
        entrypoint: 'authenticate_update_proposal',
        calldata: compiled
      };
    }
  };
}
