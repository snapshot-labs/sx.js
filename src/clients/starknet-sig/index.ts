import { AccountInterface } from 'starknet';
import fetch from 'cross-fetch';
import { domain, Propose, Vote, proposeTypes, voteTypes } from './types';

export class StarkNetSig {
  public readonly address: string;

  constructor(address: string) {
    this.address = address;
  }

  public async sign(signer: AccountInterface, address: string, message, types, primaryType) {
    const data: any = { domain, types, primaryType, message };
    const sig = await signer.signMessage(data);
    return { address, sig, data };
  }

  public async send(envelop) {
    const init = {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        jsonrpc: '2.0',
        method: 'send',
        params: { envelop },
        id: null
      })
    };
    const res = await fetch(this.address, init);
    const json = await res.json();
    return json.result;
  }

  public async propose(web3: AccountInterface, address: string, message: Propose) {
    return await this.sign(web3, address, message, proposeTypes, 'Propose');
  }

  public async vote(web3: AccountInterface, address: string, message: Vote) {
    return await this.sign(web3, address, message, voteTypes, 'Vote');
  }
}
