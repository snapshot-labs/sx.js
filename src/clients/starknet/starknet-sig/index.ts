import randomBytes from 'randombytes';
import { Account, CallData, shortString, typedData, uint256 } from 'starknet';
import { getStrategiesWithParams } from '../../../utils/strategies';
import { baseDomain, proposeTypes, updateProposalTypes, voteTypes } from './types';
import {
  ClientConfig,
  ClientOpts,
  Envelope,
  Propose,
  UpdateProposal,
  Vote,
  StarknetEIP712ProposeMessage,
  StarknetEIP712UpdateProposalMessage,
  StarknetEIP712VoteMessage,
  SignatureData
} from '../../../types';
import { defaultNetwork } from '../../..';

export class StarkNetSig {
  config: ClientConfig & { manaUrl: string };

  constructor(opts: ClientOpts & { manaUrl: string }) {
    this.config = {
      networkConfig: defaultNetwork,
      ...opts
    };
  }

  generateSalt() {
    return `0x${randomBytes(4).toString('hex')}`;
  }

  public async send(envelope) {
    const body = {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        jsonrpc: '2.0',
        method: 'send',
        params: { envelope },
        id: null
      })
    };

    const res = await fetch(
      `${this.config.manaUrl}/stark_rpc/${this.config.networkConfig.eip712ChainId}`,
      body
    );
    const json = await res.json();

    return json.result;
  }

  public async sign<
    T extends
      | StarknetEIP712ProposeMessage
      | StarknetEIP712UpdateProposalMessage
      | StarknetEIP712VoteMessage
  >(
    signer: Account,
    verifyingContract: string,
    message: T,
    types: any,
    primaryType: string
  ): Promise<SignatureData> {
    const domain = {
      ...baseDomain,
      chainId: this.config.networkConfig.eip712ChainId,
      verifyingContract
    };

    const data: typedData.TypedData = {
      types,
      primaryType,
      domain,
      message
    };

    const signature = await signer.signMessage(data);

    return {
      address: signer.address,
      signature: Array.isArray(signature)
        ? signature.map(v => `0x${BigInt(v).toString(16)}`)
        : [`0x${signature.r.toString(16)}`, `0x${signature.s.toString(16)}`],
      message,
      primaryType
    };
  }

  public async propose({
    signer,
    data
  }: {
    signer: Account;
    data: Propose;
  }): Promise<Envelope<Propose>> {
    const address = signer.address;

    const userStrategies = await getStrategiesWithParams(
      'propose',
      data.strategies,
      address,
      data,
      this.config
    );

    const message = {
      space: data.space,
      author: address,
      executionStrategy: {
        address: data.executionStrategy.addr,
        params: data.executionStrategy.params
      },
      userProposalValidationParams: CallData.compile({
        userStrategies
      }),
      metadataUri: shortString
        .splitLongString(data.metadataUri)
        .map(str => shortString.encodeShortString(str)),
      salt: this.generateSalt()
    };

    const signatureData = await this.sign(
      signer,
      data.authenticator,
      message,
      proposeTypes,
      'Propose'
    );

    return {
      signatureData,
      data
    };
  }

  public async updateProposal({
    signer,
    data
  }: {
    signer: Account;
    data: UpdateProposal;
  }): Promise<Envelope<UpdateProposal>> {
    const address = signer.address;

    const message = {
      space: data.space,
      author: address,
      proposalId: uint256.bnToUint256(data.proposal),
      executionStrategy: {
        address: data.executionStrategy.addr,
        params: data.executionStrategy.params
      },
      metadataUri: shortString
        .splitLongString(data.metadataUri)
        .map(str => shortString.encodeShortString(str)),
      salt: this.generateSalt()
    };

    const signatureData = await this.sign(
      signer,
      data.authenticator,
      message,
      updateProposalTypes,
      'UpdateProposal'
    );

    return {
      signatureData,
      data
    };
  }

  public async vote({ signer, data }: { signer: Account; data: Vote }): Promise<Envelope<Vote>> {
    const address = signer.address;

    const userVotingStrategies = await getStrategiesWithParams(
      'vote',
      data.strategies,
      address,
      data,
      this.config
    );

    const message = {
      space: data.space,
      voter: address,
      proposalId: uint256.bnToUint256(data.proposal),
      choice: `0x${data.choice.toString(16)}`,
      userVotingStrategies,
      metadataUri: shortString.splitLongString('').map(str => shortString.encodeShortString(str))
    };

    const signatureData = await this.sign(signer, data.authenticator, message, voteTypes, 'Vote');

    return {
      signatureData,
      data
    };
  }
}
