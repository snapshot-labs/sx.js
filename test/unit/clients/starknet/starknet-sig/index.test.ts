import { Account, Provider } from 'starknet';
import { StarkNetSig } from '../../../../../src/clients/starknet/starknet-sig';

describe('StarkNetSig', () => {
  const address = '0x7d2f37b75a5e779f7da01c22acee1b66c39e8ba470ee5448f05e1462afcedb4';
  const privateKey = '0xcd613e30d8f16adf91b7584a2265b1f5';

  const starkProvider = new Provider({
    sequencer: {
      baseUrl: 'http://127.0.0.1:5050'
    }
  });

  const account = new Account(starkProvider, address, privateKey);

  const client = new StarkNetSig({
    starkProvider,
    ethUrl: 'https://rpc.brovider.xyz/5',
    manaUrl: 'http://localhost:3001'
  });

  beforeAll(() => {
    jest.spyOn(client, 'generateSalt').mockImplementation(() => '0x0');
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('should create propose envelope', async () => {
    const envelope = await client.propose({
      signer: account,
      data: {
        space: '0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8',
        authenticator: '0x00e6533da3322019c3e26bd6942b647a74593af805021003bab707267717952a',
        strategies: [
          {
            address: '0x510d1e6d386a2adcfc6f2a57f80c4c4268baeccbd4a09334e843b17ce9225ee',
            index: 0
          }
        ],
        executionStrategy: {
          addr: '0x040de235a2b53e921d37c2ea2b160750ca2e94f01d709f78f870963559de8fbe',
          params: ['0x101']
        },
        metadataUri: 'ipfs://QmNrm6xKuib1THtWkiN5CKtBEerQCDpUtmgDqiaU2xDmca'
      }
    });

    expect(envelope).toMatchSnapshot();
  });

  it('should create update proposal envelope', async () => {
    const envelope = await client.updateProposal({
      signer: account,
      data: {
        space: '0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8',
        authenticator: '0x00e6533da3322019c3e26bd6942b647a74593af805021003bab707267717952a',
        executionStrategy: {
          addr: '0x040de235a2b53e921d37c2ea2b160750ca2e94f01d709f78f870963559de8fbe',
          params: ['0x101']
        },
        proposal: 1,
        metadataUri: 'ipfs://QmNrm6xKuib1THtWkiN5CKtBEerQCDpUtmgDqiaU2xDmca'
      }
    });

    expect(envelope).toMatchSnapshot();
  });

  it('should create vote envelope', async () => {
    const envelope = await client.vote({
      signer: account,
      data: {
        space: '0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8',
        authenticator: '0x00e6533da3322019c3e26bd6942b647a74593af805021003bab707267717952a',
        strategies: [
          {
            address: '0x510d1e6d386a2adcfc6f2a57f80c4c4268baeccbd4a09334e843b17ce9225ee',
            index: 0
          }
        ],
        proposal: 1,
        choice: 1
      }
    });

    expect(envelope).toMatchSnapshot();
  });
});
