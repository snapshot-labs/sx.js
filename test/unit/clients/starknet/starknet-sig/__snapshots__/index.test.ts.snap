// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StarkNetSig should create propose envelope 1`] = `
{
  "data": {
    "authenticator": "0x00e6533da3322019c3e26bd6942b647a74593af805021003bab707267717952a",
    "executionStrategy": {
      "addr": "0x040de235a2b53e921d37c2ea2b160750ca2e94f01d709f78f870963559de8fbe",
      "params": [
        "0x101",
      ],
    },
    "metadataUri": "ipfs://QmNrm6xKuib1THtWkiN5CKtBEerQCDpUtmgDqiaU2xDmca",
    "space": "0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8",
    "strategies": [
      {
        "address": "0x510d1e6d386a2adcfc6f2a57f80c4c4268baeccbd4a09334e843b17ce9225ee",
        "index": 0,
      },
    ],
  },
  "signatureData": {
    "address": "0x7d2f37b75a5e779f7da01c22acee1b66c39e8ba470ee5448f05e1462afcedb4",
    "message": {
      "author": "0x7d2f37b75a5e779f7da01c22acee1b66c39e8ba470ee5448f05e1462afcedb4",
      "executionStrategy": {
        "address": "0x040de235a2b53e921d37c2ea2b160750ca2e94f01d709f78f870963559de8fbe",
        "params": [
          "0x101",
        ],
      },
      "metadataUri": [
        "0x697066733a2f2f516d4e726d36784b75696231544874576b694e35434b7442",
        "0x4565725143447055746d6744716961553278446d6361",
      ],
      "salt": "0x0",
      "space": "0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8",
      "userProposalValidationParams": [
        "1",
        "0",
        "1",
        "0",
      ],
    },
    "primaryType": "Propose",
    "signature": [
      "0x34b23e74aaccaed17b38fadb9efed0a2778ccdb01925524e206d60e3d03544f",
      "0x36eea1b8e8258cb786c5f3f052853bd89e130651d6ab2526394123d840bd763",
    ],
  },
}
`;

exports[`StarkNetSig should create update proposal envelope 1`] = `
{
  "data": {
    "authenticator": "0x00e6533da3322019c3e26bd6942b647a74593af805021003bab707267717952a",
    "executionStrategy": {
      "addr": "0x040de235a2b53e921d37c2ea2b160750ca2e94f01d709f78f870963559de8fbe",
      "params": [
        "0x101",
      ],
    },
    "metadataUri": "ipfs://QmNrm6xKuib1THtWkiN5CKtBEerQCDpUtmgDqiaU2xDmca",
    "proposal": 1,
    "space": "0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8",
  },
  "signatureData": {
    "address": "0x7d2f37b75a5e779f7da01c22acee1b66c39e8ba470ee5448f05e1462afcedb4",
    "message": {
      "author": "0x7d2f37b75a5e779f7da01c22acee1b66c39e8ba470ee5448f05e1462afcedb4",
      "executionStrategy": {
        "address": "0x040de235a2b53e921d37c2ea2b160750ca2e94f01d709f78f870963559de8fbe",
        "params": [
          "0x101",
        ],
      },
      "metadataUri": [
        "0x697066733a2f2f516d4e726d36784b75696231544874576b694e35434b7442",
        "0x4565725143447055746d6744716961553278446d6361",
      ],
      "proposalId": {
        "high": "0x0",
        "low": "0x1",
      },
      "salt": "0x0",
      "space": "0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8",
    },
    "primaryType": "UpdateProposal",
    "signature": [
      "0x5450f74b8634b40a475612c911d6eb28a344e15e249502b4206cd31d3ba7342",
      "0x7e8346c6bad7e3ebd46fa7e990e9537feb046ed8a9da6986b6f5e7a64cf952d",
    ],
  },
}
`;

exports[`StarkNetSig should create vote envelope 1`] = `
{
  "data": {
    "authenticator": "0x00e6533da3322019c3e26bd6942b647a74593af805021003bab707267717952a",
    "choice": 1,
    "proposal": 1,
    "space": "0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8",
    "strategies": [
      {
        "address": "0x510d1e6d386a2adcfc6f2a57f80c4c4268baeccbd4a09334e843b17ce9225ee",
        "index": 0,
      },
    ],
  },
  "signatureData": {
    "address": "0x7d2f37b75a5e779f7da01c22acee1b66c39e8ba470ee5448f05e1462afcedb4",
    "message": {
      "choice": "0x1",
      "metadataUri": [],
      "proposalId": {
        "high": "0x0",
        "low": "0x1",
      },
      "space": "0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8",
      "userVotingStrategies": [
        {
          "index": 0,
          "params": [
            "0x0",
          ],
        },
      ],
      "voter": "0x7d2f37b75a5e779f7da01c22acee1b66c39e8ba470ee5448f05e1462afcedb4",
    },
    "primaryType": "Vote",
    "signature": [
      "0x6419e7b5a19568e3fd1918a33062c68de03ccac106bfc9a5c0dcb387754749c",
      "0x158b4fae20a6aa32ee232b7b067385ec2ae59f64e2c3afe30695b2775234dd",
    ],
  },
}
`;
